add --session-timeout

okken opened the pull request:
Addresses request #163

okken commented:
Adding WIP because docs need to be updated.

flub commented on pytest_timeout.py:
```diff
@@ -79,6 +85,15 @@ def pytest_addoption(parser):
+        dest="suite_timeout",
+        default=None,
+        type=float,
+        metavar="minutes",
```
`metavar` is usually upper case?

But more importantly, I don't know how I feel about minutes.  All other durations in pytest-timeout are seconds.  At least once there have been people we have wanted a more flexible duration specification that makes the unit clearer.  But that doesn't exist so I guess we're stuck with just floats (which, btw, was picked consciously originally to match the unit of the `time.time()` family of functions).  But mixing the unit of that float in different places does not seem great.

Would you mind using seconds here?  Or is the larger numbers really off-putting?

okken replied:
Seconds is fine. 
Upper case metavar does look better. thanks.

flub commented on pytest_timeout.py:
```diff
@@ -43,6 +44,11 @@
+SUITE_TIMEOUT_DESC = """
+Timeout in minutes for entire suite.  Default is None which
+means no timeout. Timeout is checked between tests, and will not interrupt a test 
+in progress. Can be specified as a float for partial minutes.
```
This seems like a reasonable design choice.

flub commented on pytest_timeout.py:
```diff
@@ -507,3 +531,9 @@ def dump_stacks(terminal):
+def pytest_runtest_logfinish(nodeid, location):
+    if _suite_expire_time and _suite_expire_time < time.time():
+        pytest.exit(f"suite-timeout: {_suite_timeout_minutes} minutes exceeded",
+                    returncode=0)
```
This should not be a successful exit should it be?  Is `returncode=0` correct?

flub replied:
Using this hook seems like it might ineterfere with reporting of the last test, especially since you don't use trylast or something.

Have you considered doing this in the existing `pytest_runtest_protocol` hook?  There you have access to the `session` (via `item`) and can use `session.shouldfail` which is documented (in `pytest_runtestloop` docs) to be a way to terminate the suite.

Additionally since item gives you a great deal of access, you can also get rid of the globals by storing the session timeout directly in the config.  I'm not too keen on those globals.

I'm not saying what I suggest here is right, but it needs experimentation and consideration of what the tradeoffs are to achieve this.

okken replied:
I like the idea of using `pytest_runtest_protocol`, but it's difficult.  
Setting `session.shouldstop` or `session.shouldfail` makes a lot of sense. (not sure which one though).
So I want to do that.

However, I don't know how to add code "at the end" of `pytest_runtest_protocol`.
If I add the timeout using this hook, it then completes the protocol for that item. 
So, at the end of a test, if we've exceeded the timeout, we don't know it until the runtest_protocol for the next test. 
Then we find out and finish. 
So we end up running an extra test after the timeout.

The closest thing I can figure out is to use `make_report` and `shouldstop` (or `shouldfail`).
```
def pytest_runtest_makereport(item, call):
    session = item.session
    if _expire_time and _expire_time < time.time():
        session.shouldstop = f"suite-timeout: {_suite_timeout} sec exceeded"
```

I think this is probably the best of both worlds with the hooks available, and doesn't use `exit()` which is rather ugly. I actually just didn't know about `shouldfail` and `shouldstop`. Thanks for pointing those out.

okken replied:
ooo. my first use of StashKey, fun.

flub commented on test_pytest_timeout.py:
```diff
@@ -603,3 +603,22 @@ def test_foo():
+    """
+    )
+    # each parametrization runs for 0.1 sec
+    # or about 0.00166 seconds each
```
minutes?  but why make the maths complicated?  why not sleep for 1 second?

Trying to make this test suite fast just makes it flaky, make the timeouts used comfortably large, never smaller than 1s.  So I'd aim for a "suite timeout" of 1 second here.

okken replied:
Yea. Seconds is more consistent with the rest of the timeouts.

okken replied:
And letting the test suite run for a second seems ok.

okken commented:
This is awesome feedback. 
thanks so much for taking the time to add detailed notes. I’m grateful for your perspective. 
I will play with this a bit and try some changes.

okken commented:
@flub I've made, I think, all suggested changes except the timeout location. 
I went with `pytest_runtest_makereport` because that was the last point in `pytest_runtest_protocol` list of hooks that I could get hold of `item`, and thus `session`.
If I did the work right in `pytest_runtest_protocol`, I would end up with an extra test being run after the timeout.

flub commented on pytest_timeout.py:
```diff
@@ -507,3 +534,12 @@ def dump_stacks(terminal):
         terminal.write("".join(traceback.format_stack(frame)))
+
+
+def pytest_runtest_makereport(item, call):
```
this hook is apparently called many times for a single test, see `CallInfo`. Furthermore this hook is documented to "stop at first non-None result".  are you sure this hook will always correctly interact with the hook that writes to the terminal or whatever output?

okken replied:
I'll add a test for `if call.when == "teardown"` so it's only called once per test.
I can also add a `hookimpl(tryfirst=True)` to try to ensure that other hooks that might return non-None run after this one. 
But really, I don't think there are guarantees

flub reviewed (commented):
This should probably be added to the `pytest_report_header` as well.  Or maybe only if it is set to keep the noise down

flub commented:
> If I did the work right in `pytest_runtest_protocol`, I would end up with an extra test being run after the timeout.

I'm not sure I follow this yet.  The docs I can find say:
```
If at any point session.shouldfail or session.shouldstop are set, the loop is terminated after the runtest protocol for the current item is finished.
```

Looking at `pytest_runtestloop()` in `main.py` seems to support what those docs are saying.  It calls `raise session.Failed(session.shouldfail) after the call to `pytest_runtest_protocol()` thus cleanly terminating in between two items, without starting the next one or starting to call any reporting hooks about the next item.

What am I missing?
(I'm just reading code, not running so am probably missing something...)

okken commented on pytest_timeout.py:
```diff
@@ -18,6 +19,7 @@
 
 
 __all__ = ("is_debugging", "Settings")
+suite_timeout_key = pytest.StashKey[float]()
```
Uppercase is fine with me. I was going by convention of uses of StashKey in pytest-dev: https://github.com/search?q=org%3Apytest-dev%20pytest.StashKey&type=code
However, I like it uppercase. good call.

okken commented on pytest_timeout.py:
```diff
@@ -119,9 +135,13 @@ def pytest_addhooks(pluginmanager):
     pluginmanager.add_hookspecs(TimeoutHooks)
 
 
+_suite_expire_time = 0
```
yes. oops.

okken commented on pytest_timeout.py:
```diff
@@ -119,9 +135,13 @@ def pytest_addhooks(pluginmanager):
 @pytest.hookimpl
 def pytest_configure(config):
     """Register the marker so it shows up in --markers output."""
+    global _suite_expire_time, _suite_timeout_minutes
```
yep

okken commented:
> This should probably be added to the `pytest_report_header` as well. Or maybe only if it is set to keep the noise down

good idea. I usually run with `--no-header` so it didn't occur to me. :)

okken commented:
> > If I did the work right in `pytest_runtest_protocol`, I would end up with an extra test being run after the timeout.
> 
> I'm not sure I follow this yet. The docs I can find say:
> 
> ```
> If at any point session.shouldfail or session.shouldstop are set, the loop is terminated after the runtest protocol for the current item is finished.
> ```
> 
> Looking at `pytest_runtestloop()` in `main.py` seems to support what those docs are saying. It calls `raise session.Failed(session.shouldfail) after the call to `pytest_runtest_protocol()` thus cleanly terminating in between two items, without starting the next one or starting to call any reporting hooks about the next item.
> 
> What am I missing? (I'm just reading code, not running so am probably missing something...)

What I'd really like is a hook right before `pytest_runtest_protocol` to be able to NOT run the next test if a timeout occurs.
But we don't have that. 

Example: 
10 tests that each take say 10 sec setup, 10 sec test, 10 sec teardown. So total of 30 sec per full test. 
If we set suite timeout to 15 sec. we expect one test to complete, and not run a second test.

However, if we check for timeout in `pytest_runtest_protocol` we will see a timeout right when we start the second test, and we mark it to stop testing, then we will complete 2 tests, and run for 60 seconds.

If we check timeout in `pytest_runtest_teardown`, we still only check the timeout at the beginning of the teardown. It's not terrible there. But if someone has long teardowns, and the timeout is during the teardown, then we don't see the timeout until the next test is done.

So I picked the hook that can be checked after the teardown, and that really is only `pytest_runtest_makereport`

flub commented on pytest_timeout.py:
```diff
@@ -182,15 +205,23 @@ def pytest_runtest_call(item):
+
+    suite_timeout = config.getoption("--suite-timeout")
+    if suite_timeout:
+        timeout_header.append("suite timeout: %ss" % suite_timeout)
```
can you add some tests for this?  I suspect you missed some newlines here.

flub commented:
> What I'd really like is a hook right before `pytest_runtest_protocol` to be able to NOT run the next test if a timeout occurs. But we don't have that.
> 
> Example: 10 tests that each take say 10 sec setup, 10 sec test, 10 sec teardown. So total of 30 sec per full test. If we set suite timeout to 15 sec. we expect one test to complete, and not run a second test.

Right, let's make a test for exactly this scenario so we can play with this and make sure it behaves right.  A test with two tests, and the suite-timeout is expected to trigger during the first test.

Because while I entirely agree with all your logic and desire to when to interrupt, I still don't understand what the thing is that stops the check from being added **at the end** of `pytest_runtest_protocol`.  That way it'll always run the first test in a suite, but given the design you proposed that is what I would expect.

> However, if we check for timeout in `pytest_runtest_protocol` we will see a timeout right when we start the second test, and we mark it to stop testing, then we will complete 2 tests, and run for 60 seconds.
> 
> If we check timeout in `pytest_runtest_teardown`, we still only check the timeout at the beginning of the teardown. It's not terrible there. But if someone has long teardowns, and the timeout is during the teardown, then we don't see the timeout until the next test is done.

I think when using the `pytest_runtest_teardown` hook you can also choose to run whatever code before **or** after the teardown?  It is the hook that invokes the teardown, so it can do whatever is right?  (not that I think we need to do something in this hook, but I'm trying to understand why you think this is an issue)

okken commented:
@flub Thanks for your patience with me and this PR. 
I honestly didn't understand how `@pytest.hookimpl(hookwrapper=True)` and `yield` worked together to allow code to run **at the end**.

I've moved the suite timeout check to `pytest_runtest_protocol`. It was already being used by the plugin, so I tacked on the suite timeout check at the end of the existing implementation.

**Question:** Does it make sense to call this flag `--session-timeout` instead of `--suite-timeout`?  

It just occurred to me that the rest of pytest generally uses the term "session" instead of "suite". 
Both are used in the documentation, but I think the code only uses "session".

okken commented:
I'm going to assume you agree that `--session-timeout` makes more sense, so changing the PR.
Let me know if you like `--suite-timeout` better.

flub commented:
> @flub Thanks for your patience with me and this PR. I honestly didn't understand how `@pytest.hookimpl(hookwrapper=True)` and `yield` worked together to allow code to run **at the end**.

Thanks for keeping at it!  And bearing with my slow responses (I've been ill in between as well :( )

> I've moved the suite timeout check to `pytest_runtest_protocol`. It was already being used by the plugin, so I tacked on the suite timeout check at the end of the existing implementation.

Glad to see this is working out!  I think the changes are much simpler now.

> **Question:** Does it make sense to call this flag `--session-timeout` instead of `--suite-timeout`?
> 
> It just occurred to me that the rest of pytest generally uses the term "session" instead of "suite". Both are used in the documentation, but I think the code only uses "session".

Yep, that's probably better.

----

I think there are a few more tests that would be really helpful to make sure this behaves right:

- The test for the header output.  We're concatenating strings now and we may want to really match the output to what we think it is.
- I think the test described earlier that ensures a timeout occurs exactly in between two given tests would be good.  This is what the feature promises so we should test it.

okken commented:
First, I added a test that would do the 2 test thing, with a slow fixture also, with this comment:

    # 2 tests, each with 0.5 sec timeouts
    # each using a fixture with 0.5 sec setup and 0.5 sec teardown
    # So about 1.5 seconds per test, ~3 sec total,
    # Therefore:
    # A timeout of 1.25 sec should happen during the teardown of the first test
    # and the second test should NOT be run

Then I realized that's also sufficient to test the timeout, so I just changed the one existing test to do that.

Second, I extended `test_header` to include session timeout, rather than add another header test just for session.

flub commented on test_pytest_timeout.py:
```diff
@@ -603,3 +608,32 @@ def test_foo():
+    # So about 1.5 seconds per test, ~3 sec total,
+    # Therefore:
+    # A timeout of 1.25 sec should happen during the teardown of the first test
+    # and the second test should NOT be run
```
i don't think this is asserted?  not immediately sure how you would assert it, maybe with more verbose output and more output lines to match.  or maybe there's something easier.

flub replied:
i don't think this is addressed yet?

flub commented on test_pytest_timeout.py:
```diff
@@ -603,3 +608,32 @@ def test_foo():
+    # each using a fixture with 0.5 sec setup and 0.5 sec teardown
+    # So about 1.5 seconds per test, ~3 sec total,
+    # Therefore:
+    # A timeout of 1.25 sec should happen during the teardown of the first test
```
0.25 from an edge is too brittle. experience tells me you can't really work with sub-seconds times in this test suite, it gets flaky otherwise.

Make the total test duration 2s and the session timeout 1s?

flub replied:
All the numbers in this description are now wrong.  maybe just describe that this is designed to timeout during the first test and ensure the first test still runs to completion but the 2nd test is not started?

flub reviewed (commented):
Looking good!  Maybe time to start editing the README :)

okken commented:
I'm really not sure where the right place to put the Session Timeout info in the docs, so I just:
- added a section at bottom, right before the changelog
- added a line to the 'unreleased' section of the changelog

When modifying the changelog I noticed the ini file setting for timeout, and figured session_timeout would also be good. So I added that to the code and added an inifile test for session_timeout

flub commented on README.rst:
```diff
@@ -340,6 +340,38 @@ function:
+Session Timeout
+===============
+
+The above mentioned timeouts are all per test function. You can also set a
+session timeout in seconds. The following example shows a session timeout
+of 10 minutes (600 seconds)::
```
Can you describe this a bit more?  I mean, "session timeout" is probably not something that makes sense to a user not very familiar with py.test internals.  I think this paragraph should describe what this timeout applies to, then explain why this is called a "session timeout".

okken replied:
I'll attempt a re-write, but feel free to change the prose if it's still awkward.

flub commented on README.rst:
```diff
@@ -340,6 +340,38 @@ function:
+      [pytest]
+      session_timeout = 600
+
+Friendly timeouts
```
maybe "cooperative timeouts" is a better title for this?  I feel like this is a concept (cooperative multi-tasking) that maps pretty well to how this behaves.  Still, as above I would first describe both how it behaves and why it is called as such.

okken replied:
sounds good

flub commented on README.rst:
```diff
@@ -340,6 +340,38 @@ function:
+Session timeouts are "friendly" timeouts. The plugin checks the session time at the end of
+each test function, and stops further tests from running if the session timeout is exceeded.
+
+Combining session and function
```
```suggestion
Combining session and function timeouts
```

flub commented on pytest_timeout.py:
```diff
@@ -88,6 +105,7 @@ def pytest_addoption(parser):
         type="bool",
         default=False,
     )
+    parser.addini("session_timeout", SESSION_TIMEOUT_DESC)
```
good catch

flub commented on test_pytest_timeout.py:
```diff
@@ -603,3 +608,55 @@ def test_foo():
+    result.assert_outcomes(passed=1)
+
+
+def test_ini_session_timeout(pytester):
```
thanks for adding this test

flub commented:
@okken ping, would be nice to see this merged.  i was kind of waiting for this to publish a release.  unless you think it'll take longer and i should release without waiting?

okken commented:
@flub thanks for the ping. I have so much going on and this slipped through the cracks.
I'll take a look right now.

okken commented:
I just pushed up some changes. All doc and/or comment changes

okken commented:
Feel free to modify any wording as you see fit to expedite the release

okken commented:
I'm pretty sure I addressed any concerns.

flub commented:
Thanks for the work and bearing with me!
